primitives/derive/src/socket.rs

```diff
@@ -0,0 +1,118 @@
+use std::net::SocketAddr;
+
+use proc_macro2::{Delimiter, TokenStream, TokenTree};
+use quote::quote;
+
+/// Stringify [TokenStream], without inserting any spaces in between
+fn stringify_tokens(tokens: TokenStream) -> String {
+    let mut result = String::new();
+
+    for token_tree in tokens {
+        match token_tree {
+            TokenTree::Group(g) => {
+                let inner = stringify_tokens(g.stream());
+
+                let bracketed = match g.delimiter() {
+                    Delimiter::Parenthesis => format!("({})", inner),
+                    Delimiter::Brace => format!("{{{}}}", inner),
+                    Delimiter::Bracket => format!("[{}]", inner),
+                    Delimiter::None => inner,
+                };
+
+                result.push_str(&bracketed);
+            }
+            o => result.push_str(&o.to_string()),
+        }
+    }
+
+    result
+}
+
+pub fn socket_impl(input: TokenStream) -> TokenStream {
+    let input = stringify_tokens(input);
+
+    let addr = match input.parse::<SocketAddr>() {
+        Ok(addr) => addr,
+        Err(e) => {
+            let message = format!("Failed to parse {:?}: {}", input, e);
```

Comment:
```suggestion
            let message = format!("Failed to parse `{:?}`: {}", input, e);
```

I prefer to isolate user input.

Reply:
It's already using the `Debug` formatting, but it will have double quotes not back ticks. 

Using `Display` with backticks should be fine